### PR TITLE
Add fake condition to deopt in opcode, move Deoptimize out to trigger

### PR DESCRIPTION
**Pull request by chuan9**
this function in any case to set the environment used to avoid check failed.
Add fake conditions refered to [Condition](https://code.google.com/p/chromium/codesearch#chromium/src/v8/src/x64/assembler-x64.h&l=253&cl=GROK&gsn=j&rcl=1459826222), move Deoptimize out to trigger this function in any case to set the environment used to avoid check failed.
It's difficult to assign environment when needed because assignation and check belong to different stages. So move deoptimizeIf out of scope to trigger this function always only if we control the condition in the deoptimization.

### CODE REVIEW

**Comment by chuan9** [comment]
@rakuco It seems that the issues also happen on the existing master branch. Could you help format master branch?

**Comment by rakuco** [comment]
`git cl format --dry-run`, which is being invoked, only checks the lines being changed by a patch. The only thing being inherited from an old commit is the indentation/alignment of the `Condition` enum in `assembler-ia32.h` (it was added in 2008), all the other errors are coming from this patch. You can verify that by running `git cl format .` after committing this and running `git show`.

**Comment by chuan9** [comment]
@rakuco I afraid not, I run <pre><code> git cl format . </code></pre> in the clean master, then run <pre><code>git diff</code></pre>, it shows many changes in format. for example:
<pre><code>
--- a/include/v8-profiler.h
+++ b/include/v8-profiler.h
@@ -579,7 +579,7 @@ class V8_EXPORT HeapProfiler {
   void SetRetainedObjectInfo(UniqueId id, RetainedObjectInfo* info);
 
   void StartTrackingHeapObjectsXDK(int stackDepth, bool retentions,
-      bool strict_collection = false);
+                                   bool strict_collection = false);
</code></pre>

**Comment by rakuco** [comment]
It depends on what `git cl` considers to be the base for the branch you are. This is what I get here:

```
/tmp/v8-crosswalk/v8 % git branch -v
* master 83e9f09 Merge pull request #143 from rakuco/clang-roll
/tmp/v8-crosswalk/v8 % git cl format .
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
```

If I manually run `clang-format -i include/v8-profiler.h`, then I do see many changes there.

However, what I've been trying to say is that the presubmit checks run by the trybot are _not_ complaining about those existing violations, but rather about the ones you are introducing in this pull request.

This is roughly what the trybot is doing:

```sh
git checkout --detach origin/master
git am -3 <your patch>.patch
git cl format --dry-run .  # this returns with status code 2
```

And if I do this here:

```sh
git checkout --detach origin/master
git am -3 <your patch>.patch
git cl format .  # without --dry-run
```

I see this in `git status`:

```
/tmp/v8-crosswalk/v8 % git status
HEAD detached from origin/master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   src/crankshaft/ia32/lithium-codegen-ia32.cc
        modified:   src/crankshaft/x64/lithium-codegen-x64.cc
        modified:   src/ia32/assembler-ia32.h

no changes added to commit (use "git add" and/or "git commit -a")
```

In short: yes, the existing commits we have on top of v8-crosswalk likely need to have their style adjusted, but it will not influence the trybot presubmit checks done on this pull request.

**Comment by chuan9** [comment]
@rakuco I use this PR to make a patch using command <code>git format-patch</code>, and run what trybot does:<pre><code>
git checkout --detach origin/master
git am -3 <your patch>.patch
git cl format --dry-run .  # this returns with status code 2
</code></pre>
The last command returns with status code 2.
but when I run: <pre><code>
git checkout --detach origin/master
git am -3 <your patch>.patch
git cl format .  # without --dry-run
git status
HEAD detached from v8-crosswalk/master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)
	modified:   include/v8-profiler.h
	modified:   src/api.cc
	modified:   src/arm/deoptimizer-arm.cc
	modified:   src/arm64/deoptimizer-arm64.cc
	modified:   src/bootstrapper.cc
	modified:   src/crankshaft/arm/lithium-arm.cc
	modified:   src/crankshaft/arm64/lithium-arm64.cc
	modified:   src/crankshaft/hydrogen-instructions.cc
.....
</code></pre>
The last command does check and modify other files which isn't in this PR.
Will you tell me how can I find the error in this PR?

**Comment by rakuco** [comment]
> HEAD detached from v8-crosswalk/master

This is the only thing that looks suspicious to me; like if you had several remotes and `origin` was behind another one which you are using.

* Which remotes do you have and where do they point to (`git remote -v`)?
* Does `origin/master` point to commit 83e9f095223b8c1269f0b113e059fbaba8cf2c48?

**Comment by chuan9** [comment]
@rakuco I have three remote repos and origin/master doesn't point to commit 83e9f09. I change remote repo v8-crosswalk into origin and now it can work. Thanks!

**Comment by rakuco** [comment]
> Is there any differences between these two? "git cl format" diff your patch with upstream, does upstream means specific origin repo?

Branching off v8-crosswalk instead of origin should work in general, but the whole process is somewhat tricky. You can see what `git cl` does by looking at `/path/to/depot_tools/git_cl.py` (specifically, `FetchUpstreamTuple()`).

Basically, if you have a named branch created from a remote branch (e.g. you ran `git checkout -b my-work-branch v8-crosswalk/master`), then it should be able to determine the upstream branch (`v8-crosswalk/master`) and all the commits in your branch that were added after the last common commit between `v8-crosswalk/master` and `my-work-branch`. It does so by checking the output of `git config branch.my-work-branch.merge` and `git config branch.my-work-branch.remote`.

However, if you created a detached checkout (`git checkout --detached`, which the bots use because that's what `gclient sync` does by default) or created a branch from a local branch (`git checkout -b my-work-branch master`), then those `git config` calls I mentioned don't return anything and `git cl` resorts to using `origin` and `master` by default.

**Comment by rakuco** [comment]
Ping @huningxin 
The Check crashes are unrelated (we've had them since moving to M49), and I believe they're caused by lack of either memory or CPU, which should be resolved soon.

**Comment by rakuco** [comment]
Sorry about that, but you'll need to re-push the patch, I force-pushed to master because of an M49 version update.

**Comment by rakuco** [comment]
Oh, and I also took the opportunity and fixed the style errors in the previous commits.

**Comment by chuan9** [comment]
@rakuco Thanks for your explanation. I did the following things:<pre><code>
git remote -v
chuan	https://github.com/chuan9/v8-crosswalk.git (fetch)
chuan	https://github.com/chuan9/v8-crosswalk.git (push)
upstream	https://chromium.googlesource.com/v8/v8.git (fetch)
upstream	https://chromium.googlesource.com/v8/v8.git (push)
v8-crosswalk	https://github.com/crosswalk-project/v8-crosswalk.git (fetch)
v8-crosswalk	https://github.com/crosswalk-project/v8-crosswalk.git (push)
git checkout v8-crosswalk/master
git checkout -b test
git am -3 <my patch>
git cl format .
Unable to determine default branch to diff against.
Either pass complete "git diff"-style arguments, like
  git cl upload origin/master
or verify this branch is set up to track another 
(via the --track argument to "git checkout -b ...").
</pre></code>
I use <code>git checkout -b> to clone a branch from v8-crosswalk/master and apply a patch. It should know to diff v8-crosswalk/master when I use <code>git cl format . >, but you know, it failed~ Just as a topic discussion here, FYI.

**Comment by chuan9** [comment]
@huningxin Will you review this patch in your spare time? Thanks!

**Comment by rakuco** [comment]
> git checkout v8-crosswalk/master
> git checkout -b test
> git am -3 my.patch
> git cl format .

That fails because of the first two steps: `git checkout v8-crosswalk/master` creates a detached checkout again. If I do it here I even get a big warning:

```
Note: checking out 'v8-crosswalk/master'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at 476c08e... Temporarily roll v8/tools/clang (24e8c1c -> c3b6048).
```

So when you do `git checkout -b test` it will create a local branch without any information about its remote counterpart.

If you do `git checkout -b test v8-crosswalk/master`, however, everything works since you are telling git that you want to create a local branch called `test` whose upstream is `v8-crosswalk/master`.

File: src/ia32/assembler-ia32.h [hunk 1]
```diff
@@ -189,32 +189,32 @@ typedef DoubleRegister SIMD128Register;
...
-  equal         =  4,
-  not_equal     =  5,
-  below_equal   =  6,
-  above         =  7,
-  negative      =  8,
-  positive      =  9,
-  parity_even   = 10,
-  parity_odd    = 11,
-  less          = 12,
+  no_condition = -1,
```
Comment (huningxin) [inline src/ia32/assembler-ia32.h]: Any reasons to change the coding style of above lines?

**Comment by huningxin** [comment]
I have a general question, why need to introduce the fake condition? Are there any special of SIMD ops which let existing deopt infrastructure cannot handle?

**Comment by huningxin** [comment]
@chuan9 , that's fair enough. Thanks for explanation! You may need to consider adding fake condition support for ia32 into upstream if we have valid usage.

LGTM.

**Comment by rakuco** [comment]
Merged, thanks. This should also get rid of the presubmit style check errors reported in #148
